### ISSUE TITLE
Some matplotlib backends on MacOS break time.sleep

### ISSUE THREAD

**Issue by impact27** [matplotlib/matplotlib#15314]
### Bug report

**Bug summary**

Using the `qt5agg` backend breaks `time.sleep`

**Code for reproduction**

launch python3 interpreter on macosx (I have 10.14.6):

```python
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib
>>> matplotlib.use('qt5agg')
>>> import matplotlib.pyplot as plt
>>> plt.plot([1, 2, 4])
[<matplotlib.lines.Line2D object at 0x11fcb6050>]
>>> plt.show()
>>> # Close the figure
... pass
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start) 
dt =  0.0006318092346191406
>>> # Wait 30-40 seconds
... pass
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start) 
dt =  2.9464328289031982
>>> print(matplotlib.get_backend())
Qt5Agg
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start) 
dt =  0.4902780055999756
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start) 
dt =  10.000809907913208
>>> 
```

**Actual outcome**
 `time.sleep(0.0005)` sleeps for 1-10 seconds

**Expected outcome**
 `time.sleep(0.0005)` should sleep for 0.0005 seconds

**Matplotlib version**

  * Operating system: macosx 10.14.6
  * Matplotlib version: 3.1.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: Python 3.7.4
  * Jupyter version (if applicable): N/A
  * Other libraries: 

 pip

**Comment by the-c0d3br34k3r** [comment]
This may be useful.
https://github.com/matplotlib/matplotlib/issues/1646/

**Comment by dopplershift** [comment]
I can't reproduce on conda-forge Python 3.7.3 on macOS 10.14.6.

**Comment by jklymak** [comment]
I can reproduce this on my mac.  You definitely need to wait the 30+ s for it to happen.

**Comment by impact27** [comment]
This bug significantly slow down jupyter (https://github.com/ipython/ipython/issues/11602), qtconsole (https://github.com/jupyter/qtconsole/issues/379) and spyder (https://github.com/spyder-ide/spyder/issues/2902)

**Comment by jklymak** [comment]
Ok is there a version of matplotlib w/o this bug?  And is it actually a matplotlib problem or a qt5 bug?

**Comment by jklymak** [comment]
I dunno, I can make this be slow for versions back to 2.0.  But I didn't try old PyQt5 versions, so maybe that's what's changed?

**Comment by impact27** [comment]
Interestingly, I can cause this bug in ipython for both qt5 and osx, but I can't trigger it with python and `matplotlib.use('MacOSX')`

```
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %matplotlib qt5                                                         

In [2]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  0.0007369518280029297

In [3]: # wait 30 seconds                                                       

In [4]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  4.751589059829712

In [5]:                                                                         
```

```
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.


In [1]: %matplotlib osx                                                         

In [2]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  0.0007028579711914062

In [3]: # wait 30 seconds  

In [4]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  10.001049041748047

In [5]:  
```

**Comment by impact27** [comment]
If `ipython` doesn't use qt5 somehow when doing `%matplotlib osx`, I would say this is not qt5 related?

**Comment by impact27** [comment]
Also, doing:
```
plt.figure()
plt.show()
```
Restores `time.sleep` for 30 seconds:
```
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib
>>> matplotlib.use('qt5agg')
>>> import matplotlib.pyplot as plt
>>> plt.figure()
<Figure size 1280x960 with 0 Axes>
>>> plt.show()
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006420612335205078
>>> # wait 30 sec
... pass
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006079673767089844
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.00093126296997
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000752925872803
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000834226608276
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000761270523071
>>> plt.show()
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.00068211555481
>>> plt.figure()
<Figure size 1280x960 with 0 Axes>
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  1.1500260829925537
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000846862792969
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000731229782104
>>> plt.show()
>>> # close figure
... pass
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0005297660827636719
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006067752838134766
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.000614166259765625
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006029605865478516
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006079673767089844
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  0.0006279945373535156
>>> # Wait 30 secs
... pass
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  2.643141031265259
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  7.550423860549927
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  10.000853061676025
>>> import time; start = time.time(); time.sleep(0.0005); print('dt = ', time.time() - start)
dt =  7.2128190994262695
>>> 
```

**Comment by impact27** [comment]
Same thing for the MacOSX backend in ipython:

```
Last login: Sat Sep 21 08:30:52 on ttys001
MacBook-Pro-de-Quentin:~ quentinpeter$ ipython
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %matplotlib osx                                                         

In [2]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  0.0007028579711914062

In [3]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  10.001049041748047

In [4]: import matplotlib; print(matplotlib.get_backend())                      
MacOSX

In [5]: import matplotlib.pyplot as plt; plt.figure()                           
Out[5]: <Figure size 640x480 with 0 Axes>

In [6]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  0.0006263256072998047

In [7]: import time; start = time.time(); time.sleep(0.0005); print('dt = ', tim
   ...: e.time() - start)                                                       
dt =  0.0006978511810302734

In [8]:                                                                         
```

**Comment by impact27** [comment]
I think it is a qt5 bug. I don't understand why `%matplotlib osx` triggers it in ipyconsole.
Paste this code in a py file and run it:

Be sure not to leave the app in the focus (click back on the terminal as soon as the script starts to run).

```python
#!/usr/bin/env python3

import time
# Uncomment either
from PySide2 import QtWidgets, QtCore
# from PyQt5 import QtWidgets, QtCore

app = QtWidgets.QApplication([''])
timer = QtCore.QTimer(app)
timer.setSingleShot(True)
timer.timeout.connect(lambda: app.quit())
timer.start(35000)
app.exec_()
start = time.time()
time.sleep(0.0005)
print('dt = ', time.time() - start)
```

**Comment by jklymak** [comment]
Is this just a mac/qt problem or can it be replicated on Linux?

**Comment by impact27** [comment]
This is caused by AppNap: https://developer.apple.com/library/archive/documentation/Performance/Conceptual/power_efficiency_guidelines_osx/AppNap.html

**Comment by impact27** [comment]
To solve: https://pypi.org/project/appnope/

**Comment by impact27** [comment]
This can't really be solved in matplotlib because:
 - This is not actually matplotlib related, it is just a problem that appear when a process shows a figure on macos >= 10.9.
 - Even if matplotlib was willing to disactivate AppNap, this require calling objective-c code. If the user uses a subprocess, this might lead to crashes because of Mac OS High Sierra added security to restrict multithreading (https://bugs.python.org/issue33725). This will probably be solved in python 3.8.

**Comment by impact27** [comment]
The workarounds for the user are:
use appnope:
```
import appnope
appnope.nope()
```
If you need to use appnope in a subprocess (Because of reasons), `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` before starting python.

**Comment by tacaswell** [comment]
While I am not sure I see the path to fixing this, I am glad this has a better explanation than "the inputhooks went off the rails".

Thanks for sorting this out @impact27 !

**Comment by impact27** [comment]
Closing as it was solved by ipython/ipykernel#441
